### ISSUE TITLE
Body shrinking

### ISSUE THREAD

**Issue by kangax** [mkoryak/floatThead#276]
I'm seeing table's body shrink as soon as I apply `floatThead`. Here's the original state:

![screenshot 2015-12-28 12 54 04](https://cloud.githubusercontent.com/assets/383/12022948/835a535a-ad62-11e5-8ade-2d622856938e.png)

and here's what happens after `$('table').floatThead()`:

![screenshot 2015-12-28 12 54 18](https://cloud.githubusercontent.com/assets/383/12022955/9395ba98-ad62-11e5-9b18-38e2577d6260.png)

I'm testing this on [ES6 compat table](https://kangax.github.io/compat-table/es6/). I've included the lib temporarily so you can see the outcome by running `$('table').floatThead()` in a console.

Any idea what's going on?

**Comment by mkoryak** [comment]
Hey! I opened a PR a very long time ago trying to apply my plugin to your table. The problem with your table is this invalid markup:

```html
<colgroup>
        <col>
        <col class="this-browser-col">
</colgroup>
```

the number of cols within a colgroup should reflect the number of columns in the table. my plugin uses colgroups internally to make the floated header's column widths match those of the table.  It does support  using an existing colgroup if it exists, this is what it is doing here. 

I am not sure why you have that colgroup there, but assuming its there for a good reason, you will need to update it have the same number of `col` elements as the table. 

I remember that I had other issues where some kind of tooltip was floated over the TH of the hovered column, this doesnt appear to be happening anymore, so maybe you just need to fix this one thing

**Comment by mkoryak** [comment]
I missed your comment about having the lib in there. ill play around with it.

**Comment by mkoryak** [comment]
ok, i played around with it some more. There are few more issues:

1. Rules like this wont fly:
```css
#show-obsolete:checked ~ table .obsolete, #show-unstable:checked ~ table .unstable {
    display: table-cell;
}
```

when the headers are floated, now you have 2 tables in there. The floated headers no longer live in a table next to your checkbox. I like the trick of not using any javascript here, but this will longer work. You need to add js to set a class on the body - and I say the body because the plugin copies the table's classes to the new table which contains the headers so css like this will continue to work:

```css
body.show-unstable .the-giant-table  .unstable {
   display: table-cell;
}
```
this css will apply to both tables. 

2. The table's `td.category` cells contain an incorrect number for their colspan when some cells are hidden. The browser is forgiving of colspans greater than the number of columns in the table, but my plugin cant figure out the true number of columns. A quick fix to this problem looks like this:

```js
var cols = $("tr.supertest:first>td:visible").length;
$("tr.category>td").attr('colspan', cols);
```

3. I think that the colgroup is probably a relic because removing it doesnt seem to break anything `$('colgroup').remove()`

**Comment by kangax** [comment]
Thanks for looking into this! I removed colgroup; will try to fix other issues now too.

**Comment by kangax** [comment]
Hey, so I pushed few changes under a "#float" flag — http://kangax.github.io/compat-table/es6/#float

- Removed colgroup
- Changed ~ selector to body one
- Added `floatThead` initialization on load and each time unstable/obsolete checkbox is clicked and columns are rerendered

This _almost_ works now! Check it out.

One artifact I'm still seeing is this:

![screenshot 2015-12-29 12 02 31](https://cloud.githubusercontent.com/assets/383/12038200/1c46ba88-ae24-11e5-9a8a-a3f8bc013064.png)

I'll look into it next.

**Comment by kangax** [comment]
Firefox (nightly) seems to have a slight shift:

![screenshot 2015-12-29 12 11 09](https://cloud.githubusercontent.com/assets/383/12038382/9fb5886c-ae25-11e5-9101-0468d3421bfe.png)

**Comment by mkoryak** [comment]
I think the headers not lining up issue has  to do with some styles still not applying to the floated header and applying to the main table. The plugin expects that both tables are styled identically. 

also, I suggest making this change:

```js
table.floatThead = (function(floatThead) {
    var created = false;
    var fn = function() {
        if (document.location.hash.indexOf('float') > -1) {
            
            //not the best place for it, but this still needs to happen:
            var cols = $("tr.supertest:first>td:visible").length;
            $("tr.category>td").attr('colspan', cols);
            
            if( created ) {
                return floatThead.call(this);
            } else {
                return floatThead.call(this, "reflow");
            }
        }
    };
    return fn;
})(table.floatThead);
```

i can look more at this later today

**Comment by kangax** [comment]
Ah... I actually changed it later to `floatThead.apply(this, arguments)` and was later calling `floatThead('destroy')` followed by `floatThead()`. This fixed the issue with table not expanding to the right properly but introduced some sort of a glitch with offset when the table is collapsed (no obsolete shown). Reflow is a good idea!

**Comment by kangax** [comment]
Great, I tweaked your snippet and it's now working perfectly — https://github.com/kangax/compat-table/commit/24b0982784f1270df4fae8907a77bcef73d59d10

One last thing left (I think) — sorting :)

**Comment by mkoryak** [comment]
Awesome!

you basically need to reflow every time the dom in the table is modified so that the header can re-align itself. you probably need to do this at the end of the sort function. Another way to do it would be `table.triggerHandler('reflow')`, if you like that style better.

**Comment by kangax** [comment]
Reflow doesn't seem to help after sorting. Something is definitely off.

![screenshot 2015-12-29 13 15 47](https://cloud.githubusercontent.com/assets/383/12039434/6efa12d4-ae2e-11e5-9612-6d96654f2210.png)

**Comment by mkoryak** [comment]
ok, i see whats going on here.

I set overflow: hidden on the container that has the floating table, it makes sense to do this 99% of the time, in your case, you can remove it:

```css
.floatThead-container {
   overflow: visible;
}
```

but this still looks funny when you scroll. we can fix that by passing a `top: function(){ return 10 when sorting, else 0 }` to the config for the plugin.

but I think a better fix is to simply add an empty cell above the features row (when sorting), and then none of these hacks will be required.

**Comment by mkoryak** [comment]
actually there is more here. when you sort you are doing something to all `tr`s in the table. when the header is floated, the trs in the header are no longer in your table. 

either destroy floatthead before sorting, or change

https://github.com/kangax/compat-table/blob/24b0982784f1270df4fae8907a77bcef73d59d10/master.js#L460

to

`table.detach().floatThead('getRowGroups').find('tr').each(function(i, row) {`

**Comment by mkoryak** [comment]
ok, i debugged a little and see one more change you need to make, and perhaps add a bug for me.


```js
return floatThead.call(this, {
  'headerCellSelector': 'tr:last>*:visible'
});
```

**Comment by kangax** [comment]
Phew, I think everything is finally fixed, but please double-check!

http://kangax.github.io/compat-table/es6/#float

Thanks for all the help!

**Comment by mkoryak** [comment]
No problem, thanks for the table, i use it a bunch :)

so one small problem, when you are in sort mode, and you scroll, you lose the % numbers. 

try this fix:

return floatThead.call(this, {
  'headerCellSelector': 'tr:last>*:visible',
  'top': function(){
      return $('#sort:checked').length * 15; //account for .num-features sticking above header
  }
});

**Comment by kangax** [comment]
That one is tricky because there's no background behind % numbers and so even with offset they blend into the rest of the table during scroll.

**Comment by mkoryak** [comment]
ok how about this, we move the percentages back down into the cells when the headers are floating:

```js
table.on("floatThead", function(evt, isFloating, $container){ 
  $container[isFloating ? 'addClass' : 'removeClass']("floating-header");
});
```

```css
.floatThead-container.floating-header sup.num-features { 
  top: 0;
}
```

**Comment by mkoryak** [comment]
also, i noticed that you added `jquery.floatThead._.js` to your source. I suggest using `jquery.floatThead.js` from a cdn, or at least a minified version from dist. Don't use the slim version since you don't have underscore.

**Comment by kangax** [comment]
Hm, good points! Would be super awesome if you could send a PR! :)

**Comment by mkoryak** [comment]
sure, ill see what I can do tomorrow. btw, noticed another bug, when clicking on a header, that column is not sized properly. a reflow is needed somewhere... ill look tomorrow
